src/Vanilla/VanillaConnect/VanillaConnect.php

```diff
@@ -0,0 +1,372 @@
+
+    /**
+     * Validate the authentication header and fill $this->errors if there is any error.
+     *
+     * @param array $claim JWT header.
+     */
+    private function validateRequestHeader(array $claim) {
+        if (!$this->validateHeaderFields($claim, 'request')) {
+            return;
+        }
+    }
+
+    /**
+     * Validate the authentication claim and fill $this->errors if there is any error.
+     *
+     * @param array $claim JWT claim.
+     */
+    private function validateRequestClaim(array $claim) {
+        $missingKeys = array_keys(array_diff_key(self::JWT_REQUEST_CLAIM_TEMPLATE, $claim));
+        if (count($missingKeys)) {
+            $this->errors['request_missing_claim_item'] = 'The authentication JWT claim is missing the following item(s): '.implode(', ', $missingKeys);
+            return;
+        }
+
+        if (preg_match('/^\d+\.\d+\.\d+$/', $claim['version']) !== 1) {
+            $this->errors['request_invalid_version'] = 'Invalid version.';
+            return;
+        }
+
+        if (version_compare(explode('.', self::VERSION)[0], explode('.', $claim['version'])[0]) === 1) {
+            $this->errors['request_incompatible_version'] = 'The request was issued with version '.$claim['version'].
+                ' but this library needs a client of at least version '.self::VERSION;
+            return;
+        }
+    }
+
+    /**
+     * @param $claim
+     * @param $type
+     * @return bool
+     */
+    private function validateHeaderFields($claim, $type) {
+        $missingKeys = array_keys(array_diff_key(constant(VanillaConnect::class.'::JWT_'.strtoupper($type).'_HEADER_TEMPLATE'), $claim));
+        if (count($missingKeys)) {
+            $this->errors[$type.'_missing_header_item'] = 'The '.$type.' JWT header is missing the following item(s): '.implode(', ', $missingKeys);
+            return false;
+        }
+
+        if ($claim['azp'] !== $this->clientID) {
+            $this->errors[$type.'_client_id_mismatch'] = 'The JWT was issued using a different ClientID(azp) than what was expected.';
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Validate the response claim and fill $this->errors if there is any error.
+     *
+     * @param array $claim JWT claim.
+     */
+    private function validateResponseClaim(array $claim) {
+        if (isset($claim['errors'])) {
+            $this->errors = $claim['errors'];
+            return;
+        } else {
+            $missingKeys = array_keys(array_diff_key(self::JWT_RESPONSE_CLAIM_TEMPLATE, $claim));
+            if (count($missingKeys)) {
+                $this->errors['response_missing_claim_item'] = 'The JWT claim is missing the following item(s): '.implode(', ', $missingKeys);
+                return;
+            }
+
+            if (!isset($claim['id']) || $claim['id'] === '') {
+                $this->errors['response_empty_claim_id'] = 'The JWT claim\'s field "id" is empty.';
+                return;
```

Comment:
This `return` also seems unnecessary. The `return` in the "response_missing_claim_item" block can probably also be removed. You would then be able to record if both of these errors were an issue with the response.

Reply:
I've corrected most of the returns but I've let that one there since it is a nice to have if we do another verification after it.

Reply:
Wouldn't you want both errors to be recorded, if they were present? As it stands, if I were missing keys *and* the ID was empty, I would need to fix one before I even saw the other was an issue. Is there a downside to recording them both in one pass?